### PR TITLE
chore(integ): configure tests to run in pipeline

### LINKED ISSUE #199: Enable integration tests to be run in a pipeline
Currently the integration tests are configured in a way that allows people with the aws-rfdk repository checked out locally to manually configure and run them. However, if someone wants to automate them to be run in the testing step of a pipeline, it isn't currently possible. We need some tooling around allowing that to be possible.

### Use Case

I want to be able to set a pipeline that pulls the aws-rfdk repository, runs the build and package, and then starts the integration tests with the artifacts creating from the packaging step.

### Proposed Solution

To accomplish this we need to:
1. Package the integration tests and anything they require to be used during the build/package step so they are accessible alongside the aws-rfdk package in any following pipeline steps.
1. Write a script that can be used as the entry point for the testing step that will do any necessary configuration and then execute `rfdk-integ-e2e.sh` to run the tests.
1. Make sure that if anything fails, the tests clean up after themselves.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by horsmand**
Fixes #199

Tested by setting up a pipeline with the integ tests in a testing stage and getting a successful run, and also running the tests manually, from the command line like we have been doing for previous releases, to make sure the changes didn't affect that workflow.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: integ/scripts/bash/run-e2e-from-pipeline.sh
Comment (jusiskin) [inline integ/scripts/bash/run-e2e-from-pipeline.sh]: We need to remove this since we don't want to accept Mongo SSPL license on anyone's behalf if they checkout this repository and try running this.

Instead, those building a pipeline that runs these tests should decide whether to accept the SSPL for themselves and configure this environment variable in their pipeline.

We could put a sanity check that happens here and fails fast if the environment variable is not set to accept the SSPL and display a user-friendly error message.

File: integ/components/deadline/common/scripts/bash/component_e2e.sh
Comment (jusiskin) [inline integ/components/deadline/common/scripts/bash/component_e2e.sh]: nit: the re-use of the `PRE_COMPONENT_HOOK` to happen before each phase changes the semantics of the name of the hook. Should we either:

1.  Rename the hook to something like `PRE_COMPONENT_PHASE_HOOK`?
1.  Make separate hooks for each phase?
    *   `PRE_COMPONENT_DEPLOY_HOOK`
    *   `PRE_COMPONENT_TEST_HOOK`
    *   `PRE_COMPONENT_DESTROY_HOOK`

    In the case of our pipeline, we'd point them to the same bash function to assume the role.


Same feedback applies to the other scripts where we use  the hook:
*   `integ/scripts/bash/cleanup.sh`
*   `integ/scripts/bash/tear-down.sh`
*   `integ/scripts/bash/teardown-infrastructure.sh`

File: integ/components/deadline/common/scripts/bash/component_e2e.sh
Comment (jusiskin) [inline integ/components/deadline/common/scripts/bash/component_e2e.sh]: In the original place we introduced this hook

https://github.com/aws/aws-rfdk/blob/b4bfd4ec475c3e94002ac961c3b2ce6c8abd65ee/integ/scripts/bash/rfdk-integ-e2e.sh#L79,

we call the hook function with the name of the component being tested as the first argument. We should probably make this consistent and either:

1.  supply the component name here
1.  remove the component name from being passed in `integ/scripts/bash/rfdk-integ-e2e.sh`

Same feedback applies to other places you've added calls to `PRE_COMPONENT_HOOK`

File: integ/scripts/bash/run-against-dist
Comment (jusiskin) [inline integ/scripts/bash/run-against-dist]: nit: looks like a copy/paste leftover from the kitchen sink

File: integ/components/deadline/deadline_04_workerFleetHttps/bin/deadline_04_workerFleetHttps.ts
Comment (jusiskin) [inline integ/components/deadline/deadline_04_workerFleetHttps/bin/deadline_04_workerFleetHttps.ts:1]: There is a lot of duplication from the HTTP Worker fleet tests. Is there any way we can make that code common and only vary the specifics to HTTP(S) to remove the duplication?

File: integ/jest.config.js
Comment (jusiskin) [inline integ/jest.config.js]: Does this make both the `.ts` and `.js` tests run? It looks like this pull request is trying to standardize on running only the javascript files except in this one instance. Unless I'm missing something, we can probably restrict this to `**/*.test.js`.

File: integ/pack.sh
Comment (jusiskin) [inline integ/pack.sh]: As discussed offline, there may have been concern that by adding packaging to the `integ` package, there may be side-effects on the publishing and API reference doc generation phases of our pipeline.

Adding this comment here to continue this discussion. The same feedback applies to `pkglint`.

File: integ/components/deadline/deadline_04_workerFleetHttps/lib/workerFleetHttps-testing-tier.ts
Comment (jusiskin) [inline integ/components/deadline/deadline_04_workerFleetHttps/lib/workerFleetHttps-testing-tier.ts:45]: This file is almost an exact copy of `integ/components/deadline/deadline_03_workerFleetHttp/lib/workerFleetHttp-testing-tier.ts`. The only functional change is this `testSuiteId`. Can we make this a shared file and take that as a prop?

File: integ/scripts/bash/run-against-dist
Comment (jusiskin) [inline integ/scripts/bash/run-against-dist]: Do we actually require verdaccio? Wondering if we can avoid all of that by simply installing the built packages with `npm`:

```bash
# install dependencies
npm install
# overwrite production ones with packaged ones
npm install path/to/built/package.tgz
```

**Review by jusiskin — CHANGES_REQUESTED** [review]
The main feedback is to:

*   investigate downstream impacts of packaging `integ` and `pkglint`
*   remove the duplication between the HTTP/HTTPS Worker fleet tests
*   see if we can eliminate the need for verdaccio

File: integ/scripts/bash/run-against-dist
Comment (horsmand) [reply integ/scripts/bash/run-against-dist]: I believe the first `npm install` will fail due to not finding all the required packages (i.e. pkglint). This is irrelevant now that I switched away from using packaging.

File: integ/pack.sh
Comment (horsmand) [reply integ/pack.sh]: No longer packing these packages.

File: integ/jest.config.js
Comment (horsmand) [reply integ/jest.config.js]: Since I moved away from packaging the tests I changed it back to just use ts.

File: integ/scripts/bash/run-against-dist
Comment (horsmand) [reply integ/scripts/bash/run-against-dist]: Removed this script.

File: integ/components/deadline/common/scripts/bash/component_e2e.sh
Comment (horsmand) [reply integ/components/deadline/common/scripts/bash/component_e2e.sh]: I'm not sure what the component name would need to be supplied for, so I've removed it.

File: integ/components/deadline/common/scripts/bash/component_e2e.sh
Comment (horsmand) [reply integ/components/deadline/common/scripts/bash/component_e2e.sh]: I don't see a need for a separate hook for each phase so I'll change the name as you suggested. I'm thinking about calling it `PRE_AWS_INTERACTION_HOOK` since it is meant to run before any interaction with AWS.

File: integ/scripts/bash/run-e2e-from-pipeline.sh
Comment (horsmand) [reply integ/scripts/bash/run-e2e-from-pipeline.sh]: I've moved this SSPL acceptance out of here and injected it as an env var from the pipeline side.

File: integ/components/deadline/deadline_04_workerFleetHttps/bin/deadline_04_workerFleetHttps.ts
Comment (horsmand) [reply integ/components/deadline/deadline_04_workerFleetHttps/bin/deadline_04_workerFleetHttps.ts:1]: I'm hesitant to do any refactoring to remove duplicate code at this time. 
* There are differences in the requirement for certificates that would need to be added into the HTTPS app (I had left these in my original PR but my next update will remove them from the HTTP app).
* We've got 2 copies, the rule of 3 would say to wait for another file that requires similar code. When we make these run in parallel, it might make sense to split the platforms out into their own CDK apps too to minimize the run time. I'd rather look at how to refactor once at that point than doing it now and then again later.
* It's also unclear to me if these two apps will always need to be kept in sync. If changes don't need to be applied to both of them, refactoring out the duplicate code would just make more difficult.

I don't have very strong feelings against doing this, so if you still think it's a good idea I'll do the refactoring now.

File: integ/components/deadline/common/scripts/bash/component_e2e.sh
Comment (jusiskin) [reply integ/components/deadline/common/scripts/bash/component_e2e.sh]: Sounds good. Let's go with this

File: integ/components/deadline/common/scripts/bash/component_e2e.sh
Comment (jusiskin) [reply integ/components/deadline/common/scripts/bash/component_e2e.sh]: This is fine by me - especially in light of the name change to `PRE_AWS_INTERACTION_HOOK`.

File: integ/components/deadline/deadline_04_workerFleetHttps/bin/deadline_04_workerFleetHttps.ts
Comment (jusiskin) [reply integ/components/deadline/deadline_04_workerFleetHttps/bin/deadline_04_workerFleetHttps.ts:1]: These tests will rarely require modifications so long as our API remains stable and since you mentioned we plan to revisit this code soon for parallel execution, we can defer refactoring until then. Resolved.

File: integ/components/deadline/deadline_03_workerFleetHttp/bin/deadline_03_workerFleetHttp.ts
Comment (jusiskin) [inline integ/components/deadline/deadline_03_workerFleetHttp/bin/deadline_03_workerFleetHttp.ts]: nit: to avoid manually incrementing the counter variable (`i`), the callback to `forEach` can take an additional positional argument that contains the 0-based index. See [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#Parameters).

Same feedback applies to other places we are doing this.

File: integ/components/deadline/deadline_04_workerFleetHttps/test/deadline_04_workerFleetHttps.test.ts
Comment (jusiskin) [inline integ/components/deadline/deadline_04_workerFleetHttps/test/deadline_04_workerFleetHttps.test.ts]: It probably warrants throwing an exception here since these tests are only run for HTTPS, right?

File: integ/lib/render-struct.ts
Comment (jusiskin) [inline integ/lib/render-struct.ts]: I know this gave you headaches to root-cause. Can we add a comment here so these efforts are documented? Same applies to the other occurrence in `integ/lib/testing-tier.ts`

File: integ/scripts/bash/run-e2e-automated.sh
Comment (jusiskin) [inline integ/scripts/bash/run-e2e-automated.sh]: Instead of writing to a file (`$(mktemp -d)/creds.json`), it is safer to just keep this in memory. Aside from other processes/agents being able to read the file, if the EBS volume is unencrypted and the underlying block is reused for another instance, a motivated attacker may be able to recover the credentials. The credentials are time-limited, but it is still good practice and doesn't hurt.

I believe you should be able to do this with:

```bash
export CREDS="$(AWS_STS_REGIONAL_ENDPOINTS=regional aws sts assume-role \
        --role-arn $ROLE_ARN \
        --role-session-name $ROLE_SESSION_NAME \
        --external-id $ROLE_EXTERNAL_ID)"
export AWS_ACCESS_KEY_ID="$(printenv CREDS | grep "AccessKeyId" | cut -d'"' -f 4)"
export AWS_SECRET_ACCESS_KEY="$(printenv CREDS | grep "SecretAccessKey" | cut -d'"' -f 4)"
export AWS_SESSION_TOKEN="$(printenv CREDS | grep "SessionToken" | cut -d'"' -f 4)"
# Clean up the env var
unset CREDS
```

We do this already when [configuring the database connection for the repo installer](https://github.com/aws/aws-rfdk/blob/f65591f1974f603de01fae20e3566b11c77cffa9/packages/aws-rfdk/lib/deadline/lib/database-connection.ts#L184-L187). The `printenv` is used to avoid having the credentials logged in shell history and observable in procfs.

**Review by jusiskin — CHANGES_REQUESTED** [review]
Looking great. Just a few minor things left.

File: integ/scripts/bash/run-e2e-automated.sh
Comment (horsmand) [reply integ/scripts/bash/run-e2e-automated.sh]: Updated

File: integ/lib/render-struct.ts
Comment (horsmand) [reply integ/lib/render-struct.ts]: Added

File: integ/components/deadline/deadline_04_workerFleetHttps/test/deadline_04_workerFleetHttps.test.ts
Comment (horsmand) [reply integ/components/deadline/deadline_04_workerFleetHttps/test/deadline_04_workerFleetHttps.test.ts]: Yeah, makes sense, will change it.

File: integ/components/deadline/deadline_03_workerFleetHttp/bin/deadline_03_workerFleetHttp.ts
Comment (horsmand) [reply integ/components/deadline/deadline_03_workerFleetHttp/bin/deadline_03_workerFleetHttp.ts]: Fixed.

File: integ/components/deadline/deadline_04_workerFleetHttps/test/deadline_04_workerFleetHttps.test.ts
Comment (jusiskin) [reply integ/components/deadline/deadline_04_workerFleetHttps/test/deadline_04_workerFleetHttps.test.ts]: Looks good

File: integ/lib/render-struct.ts
Comment (jusiskin) [reply integ/lib/render-struct.ts]: Thanks!

**Review by jusiskin — APPROVED** [review]
Great work! Approved!

File: integ/scripts/bash/run-e2e-automated.sh
Comment (ddneilson) [inline integ/scripts/bash/run-e2e-automated.sh:35]: Could/should we always be using the latest version of Deadline? ex: If we set this to "10.1" then staging and such should pull the latest Deadline.

File: integ/scripts/bash/stage-deadline.sh
Comment (ddneilson) [inline integ/scripts/bash/stage-deadline.sh:11]: Could we use the newer arg form of stage-deadline here that doesn't require the full URIs? Or is there a use-case for this code that requires using unreleased versions of Deadline?

File: integ/scripts/bash/run-e2e-automated.sh
Comment (jusiskin) [reply integ/scripts/bash/run-e2e-automated.sh:35]: This is being tackled in #160. It looks like this one is lined up to be merged first. Once this is merged lets make a comment on that one to make sure the change is applied here.

File: integ/scripts/bash/stage-deadline.sh
Comment (jusiskin) [reply integ/scripts/bash/stage-deadline.sh:11]: Also part of #160.

EDIT: This one is easy to apply here as well though without any conflict.

**Review by ddneilson — APPROVED** [review]

**Pull request merged**
